### PR TITLE
Feature: stronger fp typing implementation throughout NRPY

### PR DESCRIPTION
**Pull request by samueltootle**
This PR is dependent on the outcome of PR https://github.com/nrpy/nrpy/pull/26

In this PR, a large portion of NRPy is refactored to leverage the ability to have variable precision based on `fp_type` and includes an updated `two_blackholes_collide` example to leverage this new capability.  For `fp_type = "double"` (which is the default), no changes have been found for the code generated by NRPy, e.g. the code generated by Baikal is the same after this PR.

In few cases new code has been introduced.  

- For example, relative difference for boundary points `EPS_REL` needs to be updated based on `fp_type` (see e.g. `nrpy/infrastructures/BHaH/CurviBoundaryConditions/CurviBoundaryConditions.py`) - the same goes for parity.
- Areas where `sp.ccode` is called directly outside of `ccodegen` requires an additional dependency on the `sp_ast` module and well as `ccodegen` dictionaries relating `fp_type` to the `sympy` types. See e.g. `setup_Cfunction_FD1_arbitrary_upwind`

For a test case, I utilized the updated `two_blackholes_collide` example which shows agreement between `fp_type = 'float'` and `fp_type = 'double'` to single precision as expected.  For the executable generated using `fp_type = 'float'`, the calculation was roughly 30% faster than double precision.

### CODE REVIEW

**Comment by samueltootle** [comment]
Resolved static analysis issues and codegen (locally).  There seems to be another race condition in github actions

File: nrpy/c_codegen.py [hunk 1]
```diff
@@ -32,14 +33,30 @@
...
 
+fp_type_to_sympy_type = {
+    # Traditional C types
+    "double": sp_ast.float64,
+    "float": sp_ast.float32,
+    "long double": sp_ast.float80,
+    # Standard C++ types
+    "std::float32_t": sp_ast.float32,
+    "std::float64_t": sp_ast.float64,
+    # Unsupported types by sympy ccode generator
```
Comment (zachetienne) [inline nrpy/c_codegen.py:53]: Let's remove this dead code -- no need to keep it around.

File: nrpy/c_codegen.py [hunk 2]
```diff
@@ -650,103 +676,129 @@ def c_codegen(
 #    codegen to output our desired fabs().
 nrpyAbs = sp.Function("nrpyAbs")
 custom_functions_for_SymPy_ccode = {
```
Comment (zachetienne) [inline nrpy/c_codegen.py:678]: Let's move this dictionary definition into helpers/custom_c_codegen_functions.py .

File: nrpy/infrastructures/BHaH/cmdline_input_and_parfiles.py [hunk 3]
```diff
@@ -287,7 +287,7 @@ def register_CFunction_cmdline_input_and_parfile_parser(
 """
     if found_REAL:
         prefunc += r"""
-static void read_double(const char *value, double *result, const char *param_name) {
+static void read_double(const char *value, REAL *result, const char *param_name) {
```
Comment (zachetienne) [inline nrpy/infrastructures/BHaH/cmdline_input_and_parfiles.py:290]: Function should be renamed read_REAL() or something like that...

Comment (samueltootle) [reply nrpy/infrastructures/BHaH/cmdline_input_and_parfiles.py]: Resolved in 902595f4e8521eeb2810c499ff088c083d5c0fc0

File: nrpy/c_codegen.py (hunk 2 above)
Comment (samueltootle) [reply nrpy/c_codegen.py]: Resolved in e53593abe7ed8555382031859042a683b3cb35d9

File: nrpy/c_codegen.py (hunk 1 above)
Comment (samueltootle) [reply nrpy/c_codegen.py:48]: Resolved in cd817074a1c5cb5c5c444242a2d2a368da6391dc

**Pull request merged**